File: backend/src/modules/producto/dto/publicar-producto.dto.ts
import { IsString, IsNumber, IsOptional, Min, MinLength, MaxLength, IsIn } from 'class-validator';
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import { TALLES_VALIDOS, TalleValido } from '../talles.constant';

export class PublicarProductoDto {
  @ApiProperty({ description: 'Nombre del producto (mínimo 3 caracteres)', example: 'Remera Algodón' })
  @IsString()
  @MinLength(3, { message: 'El nombre debe tener al menos 3 caracteres' })
  @MaxLength(255)
  readonly nombre: string;

  @ApiPropertyOptional({ description: 'Descripción del producto', example: 'Remera de algodón 100%' })
  @IsOptional()
  @IsString()
  readonly descripcion?: string | undefined;

  @ApiProperty({
    description: 'Talle del producto',
    example: 'M',
    enum: TALLES_VALIDOS,
  })
  @IsIn(TALLES_VALIDOS, { message: 'Seleccione un talle valido' })
  readonly talle: TalleValido;

  @ApiProperty({ description: 'Precio en centavos (mayor a cero)', example: 15000 })
  @Type(() => Number)
  @IsNumber()
  @Min(1, { message: 'El precio debe ser mayor a cero' })
  readonly precioCentavos: number;

  @ApiPropertyOptional({ description: 'Stock inicial', example: 100 })
  @IsOptional()
  @IsNumber()
  @Min(0)
  readonly stock?: number;

  constructor(
    nombre: string,
    talle: TalleValido,
    precioCentavos: number,
    descripcion?: string | undefined,
  ) {
    this.nombre = nombre;
    this.talle = talle;
    this.precioCentavos = precioCentavos;
    this.descripcion = descripcion;
  }
}

File: backend/src/main/java/com/sistemapedidos/modules/producto/dto/PublicarProductoDto.java
package com.sistemapedidos.modules.producto.dto;

import com.sistemapedidos.modules.producto.TallesConstants;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;

public record PublicarProductoDto(
        @NotBlank(message = "El nombre es obligatorio")
        @Size(min = 3, message = "El nombre debe tener al menos 3 caracteres")
        @Size(max = 255) String nombre,
        String descripcion,
        @NotBlank(message = "El talle es obligatorio")
        @Pattern(regexp = TallesConstants.TALLES_REGEX, message = "Seleccione un talle valido") String talle,
        @NotNull(message = "El precio es obligatorio")
        @Min(value = 1, message = "El precio debe ser mayor a cero") Long precioCentavos,
        @Min(value = 0) Integer stock) {
}
